src/big-pit/class-big-speculative-pit.php

```diff
@@ -0,0 +1,195 @@
+<?php
+/**
+ * Preloaded_Client class file
+ *
+ * @package wp-big-pit
+ */
+
+namespace Alley\WP\Big_Pit;
+
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * Track the keys fetched during a given request and preload those values the next time the same request occurs.
+ */
+final class Big_Speculative_Pit implements Client {
+	/**
+	 * Previously saved keys by group.
+	 *
+	 * @phpstan-var array<string, array<string>>
+	 *
+	 * @var array[]
+	 */
+	private array $saved_keys = [];
+
+	/**
+	 * Fetched keys by group.
+	 *
+	 * @phpstan-var array<string, array<string>>
+	 *
+	 * @var array[]
+	 */
+	private array $fetched_keys = [];
+
+	/**
+	 * Preloaded items.
+	 *
+	 * @var Items
+	 */
+	private readonly Items $items;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param Request $request Current request.
+	 * @param Client  $origin  Client instance.
+	 */
+	public function __construct(
+		private readonly Request $request,
+		private readonly Client $origin,
+	) {
+		$this->items = new Items();
+	}
+
+	/**
+	 * Boot the feature.
+	 */
+	public function boot(): void {
+		// Since this class isn't responsible for creating the table, make sure upstream classes can do so.
+		$this->origin->boot();
+
+		/*
+		 * Preload once here and be done with it, rather than having to make sure in every method that values were
+		 * preloaded and that preloading happens only once. Trades a DB query for a simpler implementation.
+		 */
+		$this->preload();
+
+		// Priority 0 so that it's visible in Query Monitor.
+		add_action( 'shutdown', [ $this, 'on_shutdown' ], 0 );
+	}
+
+	/**
+	 * Get a value.
+	 *
+	 * @param string $key   Item key.
+	 * @param string $group Item group.
+	 * @return mixed|null
+	 */
+	public function get( string $key, string $group ): mixed {
+		$this->fetched_keys[ $group ][] = $key;
```

Comment:
can the same key be "gotten" twice? Will this introduce a duplicate entry? Is that a problem?

Reply:
never mind. I see the `array_unique` further down